File: src/main/scala/robosky/ether/block/machine/infuser/AegisaltRecipe.scala
package robosky.ether.block.machine.infuser

import com.google.gson.{JsonObject, JsonParseException}
import net.fabricmc.api.{EnvType, Environment}
import net.minecraft.inventory.Inventory
import net.minecraft.item.ItemStack
import net.minecraft.recipe._
import net.minecraft.util.{DefaultedList, Identifier, JsonHelper, PacketByteBuf}
import net.minecraft.world.World
import robosky.ether.block.machine.{MachineRegistry, RecipeRegistry}

object AegisaltRecipe {

  object Serializer extends RecipeSerializer[AegisaltRecipe] {

    override def read(identifier_1: Identifier, jsonObject_1: JsonObject): AegisaltRecipe = {
      val group = JsonHelper.getString(jsonObject_1, "group")
      val array = JsonHelper.getArray(jsonObject_1, "ingredients")
      val ingredients = for {
        i <- 0 until array.size()
        ingredient_1 = Ingredient.fromJson(
          array.get(i)
        )
        if !ingredient_1.isEmpty
      } yield ingredient_1
      if (ingredients.size != 2)
        throw new JsonParseException(
          "Wrong number of ingredients for aegisalt recipe!"
        )
      else {
        val output = ShapedRecipe.getItemStack(
          JsonHelper.getObject(jsonObject_1, "result")
        )
        new AegisaltRecipe(identifier_1, group, ingredients, output)
      }
    }

    override def read(id: Identifier, buf: PacketByteBuf): AegisaltRecipe = {
      val group = buf.readString(32767)
      val ingredients = for (_ <- 0 to 1) yield Ingredient.fromPacket(buf)
      val output = buf.readItemStack()
      new AegisaltRecipe(id, group, ingredients, output)
    }

    override def write(var1: PacketByteBuf, var2: AegisaltRecipe): Unit = {
      var1.writeString(var2.getGroup)
      var2.ingredients.foreach(_.write(var1))
      var1.writeItemStack(var2.getOutput)
    }
  }

}

class AegisaltRecipe(id: Identifier, group: String, val ingredients: IndexedSeq[Ingredient], output: ItemStack)
  extends Recipe[Inventory] {
  override def matches(var1: Inventory, var2: World): Boolean =
    (0 to 1).forall(i => ingredients(i).method_8093(var1.getInvStack(i)))

  override def craft(var1: Inventory): ItemStack = output.copy()

  override def fits(var1: Int, var2: Int): Boolean = true

  override def getOutput: ItemStack = output

  override def getId: Identifier = id

  override def getSerializer: RecipeSerializer[_] = AegisaltRecipe.Serializer

  override def getType: RecipeType[_] = RecipeRegistry.aegisaltRecipe

  override def getRemainingStacks(
    inventory_1: Inventory
  ): DefaultedList[ItemStack] = super.getRemainingStacks(inventory_1)

  override def getPreviewInputs: DefaultedList[Ingredient] = {
    val l = DefaultedList.of[Ingredient]
    ingredients.foreach(l.add)
    l
  }

  override def isIgnoredInRecipeBook: Boolean = true

  @Environment(EnvType.CLIENT)
  override def getGroup: String = group

  @Environment(EnvType.CLIENT)
  override def getRecipeKindIcon: ItemStack =
    new ItemStack(MachineRegistry.aegisaltInfuser.machine.b)
}

File: src/main/java/robosky/uplands/block/machine/infuser/AegisaltRecipe.java
package robosky.uplands.block.machine.infuser;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import robosky.uplands.block.machine.MachineRegistry;
import robosky.uplands.block.machine.RecipeRegistry;

import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;
import net.minecraft.recipe.Ingredient;
import net.minecraft.recipe.Recipe;
import net.minecraft.recipe.RecipeSerializer;
import net.minecraft.recipe.RecipeType;
import net.minecraft.recipe.ShapedRecipe;
import net.minecraft.util.DefaultedList;
import net.minecraft.util.Identifier;
import net.minecraft.util.JsonHelper;
import net.minecraft.util.PacketByteBuf;
import net.minecraft.world.World;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;

public final class AegisaltRecipe implements Recipe<Inventory> {

    private final Identifier id;
    private final String group;
    private final List<Ingredient> ingredients;
    private final ItemStack output;

    public AegisaltRecipe(Identifier id, String group, List<Ingredient> ingredients, ItemStack output) {
        this.id = id;
        this.group = group;
        this.ingredients = ingredients;
        this.output = output;
    }

    public List<Ingredient> getIngredients() {
        return ingredients;
    }

    @Override
    public boolean matches(Inventory inv, World world) {
        ItemStack one = inv.getInvStack(0);
        ItemStack two = inv.getInvStack(1);
        return ingredientsMatch(one, two) || ingredientsMatch(two, one);
    }

    private boolean ingredientsMatch(ItemStack one, ItemStack two) {
        return ingredients.get(0).test(one) && ingredients.get(1).test(two);
    }

    @Override
    public ItemStack craft(Inventory inv) {
        return output.copy();
    }

    @Override
    public boolean fits(int i, int j) {
        return true;
    }

    @Override
    public ItemStack getOutput() {
        return output;
    }

    @Override
    public Identifier getId() {
        return id;
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return Serializer.INSTANCE;
    }

    @Override
    public RecipeType<?> getType() {
        return RecipeRegistry.AEGISALT_INFUSION;
    }

    @Override
    public DefaultedList<Ingredient> getPreviewInputs() {
        DefaultedList<Ingredient> l = DefaultedList.of();
        l.addAll(ingredients);
        return l;
    }

    @Override
    public boolean isIgnoredInRecipeBook() {
        return true;
    }

    @Override
    public String getGroup() {
        return group;
    }

    @Environment(EnvType.CLIENT)
    @Override
    public ItemStack getRecipeKindIcon() {
        return new ItemStack(MachineRegistry.AEGISALT_INFUSER.block);
    }

    public static class Serializer implements RecipeSerializer<AegisaltRecipe> {

        public static final Serializer INSTANCE = new Serializer();

        @Override
        public AegisaltRecipe read(Identifier id, JsonObject json) {
            String group = JsonHelper.getString(json, "group");
            JsonArray array = JsonHelper.getArray(json, "ingredients");
            List<Ingredient> ingredients = new ArrayList<>(2);
            for(JsonElement elem : array) {
                Ingredient ingredient = Ingredient.fromJson(elem);
                if(!ingredient.isEmpty()) {
                    ingredients.add(ingredient);
                }
            }
            if(ingredients.size() != 2) {
                throw new JsonParseException("Wrong number of ingredients for aegisalt recipe!");
            } else {
                ItemStack output = ShapedRecipe.getItemStack(JsonHelper.getObject(json, "result"));
                return new AegisaltRecipe(id, group, ingredients, output);
            }
        }

        @Override
        public AegisaltRecipe read(Identifier id, PacketByteBuf buf) {
            String group = buf.readString(32767);
            List<Ingredient> ingredients = Arrays.asList(Ingredient.fromPacket(buf), Ingredient.fromPacket(buf));
            ItemStack output = buf.readItemStack();
            return new AegisaltRecipe(id, group, ingredients, output);
        }

        @Override
        public void write(PacketByteBuf buf, AegisaltRecipe recipe) {
            buf.writeString(recipe.getGroup());
            recipe.ingredients.forEach(e -> e.write(buf));
            buf.writeItemStack(recipe.getOutput());
        }
    }
}
